native/src/main/java/io/ballerina/lib/data/xmldata/utils/DataUtils.java

```diff
@@ -230,33 +228,18 @@ public static Object convertStringToExpType(BString value, Type expType) {
         return result;
     }
 
-    public static void validateRequiredFields(BMap<BString, Object> currentMapValue, XmlAnalyzerData analyzerData) {
-        Map<QualifiedName, Field> fields = analyzerData.fieldHierarchy.peek().getMembers();
-        for (QualifiedName key : fields.keySet()) {
-            // Validate required array size
-            Field field = fields.get(key);
-            String fieldName = field.getFieldName();
-            if (field.getFieldType().getTag() == TypeTags.ARRAY_TAG) {
-                ArrayType arrayType = (ArrayType) field.getFieldType();
-                if (arrayType.getSize() != -1
-                        && arrayType.getSize() != ((BArray) currentMapValue.get(
-                        StringUtils.fromString(fieldName))).getLength()) {
-                    throw DiagnosticLog.error(DiagnosticErrorCode.ARRAY_SIZE_MISMATCH);
-                }
-            }
-
-            if (!SymbolFlags.isFlagOn(field.getFlags(), SymbolFlags.OPTIONAL)
-                    && !currentMapValue.containsKey(StringUtils.fromString(fieldName))) {
-                throw DiagnosticLog.error(DiagnosticErrorCode.REQUIRED_FIELD_NOT_PRESENT, fieldName);
+    public static void validateRequiredFields(XmlAnalyzerData analyzerData) {
+        Map<QualifiedName, Field> remainingFields = analyzerData.fieldHierarchy.peek().getMembers();
+        for (Field field : remainingFields.values()) {
+            if (SymbolFlags.isFlagOn(field.getFlags(), SymbolFlags.REQUIRED)) {
+                throw DiagnosticLog.error(DiagnosticErrorCode.REQUIRED_FIELD_NOT_PRESENT, field.getFieldName());
             }
         }
 
-        Map<QualifiedName, Field> attributes = analyzerData.attributeHierarchy.peek().getMembers();
-        for (QualifiedName key : attributes.keySet()) {
-            Field field = attributes.get(key);
-            String fieldName = field.getFieldName();
-            if (!SymbolFlags.isFlagOn(field.getFlags(), SymbolFlags.OPTIONAL)) {
-                throw DiagnosticLog.error(DiagnosticErrorCode.REQUIRED_ATTRIBUTE_NOT_PRESENT, fieldName);
+        Map<QualifiedName, Field> remainingAttributes = analyzerData.attributeHierarchy.peek().getMembers();
+        for (Field attribute : remainingAttributes.values()) {
+            if (!SymbolFlags.isFlagOn(attribute.getFlags(), SymbolFlags.OPTIONAL)) {
```

Comment:
What is the need of this logic?

I guess we only need to check the remaining fields are required or not?
isnt it?

Reply:
If the remaining field is optional field then no need to log the error.